v2/internal/frontend/devserver/devserver.go

```diff
@@ -122,32 +122,40 @@ func (d *DevWebServer) Run(ctx context.Context) error {
 	})
 
 	if devServerAddr := d.devServerAddr; devServerAddr != "" {
+		wg := sync.WaitGroup{}
+		wg.Add(1)
 		// Start server
 		go func(server *echo.Echo, log *logger.Logger) {
+			defer wg.Done()
 			err := server.Start(devServerAddr)
 			if err != nil {
 				log.Error(err.Error())
 			}
-			d.LogDebug("Shutdown completed")
 		}(d.server, d.logger)
 
 		d.LogDebug("Serving DevServer at http://%s", devServerAddr)
 
-		defer func() {
+		go func() {
+			wg.Wait()
+			d.LogDebug("Starting shutdown")
 			err := d.server.Shutdown(context.Background())
 			if err != nil {
 				d.logger.Error(err.Error())
 			}
+			d.LogDebug("Shutdown completed")
 		}()
```

Comment:
For consistency, perhaps we should pass everything in, like we do for startup?

Reply:
Makes sense to me

Reply:
I added the other parameters but I left the `d.LogDebug` statements the way they are.

I tried to come up with ways to avoid it, but they all felt contrived and/or painful to maintain. Let me you what you think!

Reply:
Nice one! I'm now wondering if these have to be 2 goroutines. I'm assuming the `server.Start` call is blocking otherwise the shutdown process would begin straight away. What do you think about removing the second goroutine and just put everything in the first? Then we could remove the waitgroup altogether?

Reply:
Maybe I'm missing something but aren't we now blocking the call to `server.Shutdown`? `sever.Start` blocks until someone calls a `server.Shutdown`, the WaitGroup onto which the `server.Shutdown` waits for is only decremented when `server.Start` finishes. So it seems like both are blocking each other.

Reply:
I think you're right - I missed the `defer` in the original whilst looking at github mobile 😬 I take it the dev server can be shut down with out of band actions such as `ctrl-c`?

Reply:
@stffabi Great point! I believe you're right. The problem I ran into was that the deferred `server.Shutdown()` was running immediately because `frontend.Frontend.Run()` no longer blocks.

It didn't click for me that by the time the waitgroup unblocks, the server has already shut down. And yes, `Ctrl-C` shuts down the server as expected with the current commit, I've tested it.

It may not make sense to explicitly call `server.Shutdown` anymore; though I could phase out the waitgroup and make the goroutine wait for a context cancellation instead. I'm not sure what's better, so please let me know your thoughts.

Reply:
Personally I'm fine with completely removing the shutdown for the time being. For the devserver an ungraceful shutdown is not much of a problem.
We could revisit this for the server/hybrid mode where a graceful shutdown might be more relevant.
@leaanthony what do you think?

Reply:
Yeah happy with that.